File: Problems/0300_0399/0380_Insert_Delete_GetRandom_O_1/Project_Go/solution/randomizedset.go
package solution

import (
	"math/rand"
	"time"
)

// 167ms - 177ms
type RandomizedSet struct {
	Data    []int
	DataMap map[int]int
}

func Constructor() RandomizedSet {
	rand.Seed(time.Now().UnixNano())
	return RandomizedSet{make([]int, 0), make(map[int]int, 0)}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.DataMap[val]; ok {

	} else {
		this.Data = append(this.Data, val)
		this.DataMap[val] = len(this.Data) - 1
		return true
	}
	return false
}

func (this *RandomizedSet) Remove(val int) bool {
	if index, ok := this.DataMap[val]; ok {
		delete(this.DataMap, val)
		lastVal := this.Data[len(this.Data)-1]
		this.Data = this.Data[:len(this.Data)-1]
		if lastVal != val {
			this.Data[index] = lastVal
			this.DataMap[lastVal] = index
		}
		return true
	}
	return false
}

func (this *RandomizedSet) GetRandom() int {
	return this.Data[rand.Intn(len(this.Data))]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: Problems/0300_0399/0380_Insert_Delete_GetRandom_O_1/Project_Java/RandomizedSet.java
import java.util.*;

// 21ms - 26ms
class RandomizedSet {
    HashMap<Integer, Integer> data_map;
    List<Integer> data;
    Random rand;

    public RandomizedSet() {
        data_map = new HashMap<>();
        data = new ArrayList<>();
        rand = new Random();
    }
    
    public boolean insert(int val) {
        if (!data_map.containsKey(val)) {
            data.add(val);
            data_map.put(val, data.size() - 1);
            return true;
        }
        return false;
    }
    
    public boolean remove(int val) {
        if (data_map.containsKey(val)) {
            int index = data_map.remove(val);
            int lastVal = data.remove(data.size() - 1);
            if (lastVal != val) {
                data.set(index, lastVal);
                data_map.put(lastVal, index);
            }
            return true;
        }
        return false;
    }
    
    public int getRandom() {
        return data.get(rand.nextInt(data.size()));
    }
}

/*
public int[] shuffle() {
    Random rand = new Random();
    int[] ans = arr.clone();
    for (int i = 0; i < ans.length; i++) {
        int swp_num = rand.nextInt(ans.length);
        int temp = ans[i];
        ans[i] = ans[swp_num];
        ans[swp_num] = temp;
    }
    return ans;
}
*/

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */
 